walletnode/services/artworksearch/thumbnail/thumbnail.go

```diff
@@ -0,0 +1,163 @@
+type Helper interface {
+	Connect(ctx context.Context, connections uint) error
+	Fetch(ctx context.Context, key string) ([]byte, error)
+	Close()
+}
+
+type thumbnailHelper struct {
+	pastelClient pastel.Client
+	nodeClient   nodeClient.Client
+
+	reqCh      chan request
+	timeOut    time.Duration
+	isClosed   bool
+	closeMutex sync.Mutex
+}
+
+// New returns a new instance of thumbnailHelper as Helper
+func New(pastelClient pastel.Client, nodeClient nodeClient.Client, timeout time.Duration) Helper {
+	return &thumbnailHelper{
+		pastelClient: pastelClient,
+		nodeClient:   nodeClient,
+		timeOut:      timeout,
+		reqCh:        make(chan request),
+	}
+}
+
+type response struct {
+	data []byte
+	err  error
+}
+
+type request struct {
+	key    string
+	respCh chan *response
+}
+
+// Connect creates `conncetions` no. of connections with supernode & starts thumbnail request listeners
+func (t *thumbnailHelper) Connect(ctx context.Context, connections uint) error {
+	if connections > maxConnections {
+		return errors.New(maxConnectionsErr)
+	}
+
+	nodes, err := t.pastelTopNodes(ctx, connections)
+	if err != nil {
+		return fmt.Errorf("pastelTopNodes: %v", err)
+	}
+
+	for _, node := range nodes {
+		node := node
+		if err := node.Connect(ctx, t.timeOut); err != nil {
+			return err
+		}
+
+		go func() {
+			t.listen(ctx, node)
+			t.Close()
+			if err := node.Connection.Close(); err != nil {
+				log.WithContext(ctx).WithError(err).Error("ThumbnailPuller.Start-Close node connection")
+			}
+		}()
+	}
+
+	return nil
+}
+
+func (t *thumbnailHelper) listen(ctx context.Context, n *node.Node) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case req, ok := <-t.reqCh:
+			if !ok {
+				return
+			}
+
+			// TODO: get thumbnail from supernode using Node
+			// hard-coding fetched thumbnail for dev purposes
+			fmt.Println("thumb-key:", req.key, n.PastelID())
+			thumb := []byte{}
+			resp := &response{data: thumb}
+			req.respCh <- resp
+		}
+	}
+}
+
+// Fetch gets thumbnail
+func (t *thumbnailHelper) Fetch(ctx context.Context, key string) (data []byte, err error) {
+	if t.IsClosed() {
+		return data, errors.New("connection is closed")
+	}
+
+	respCh := make(chan *response)
+	req := request{key: key, respCh: respCh}
+
+	go func() {
+		t.reqCh <- req
+	}()
+
+	select {
+	case <-ctx.Done():
+		return data, nil
+	case resp := <-respCh:
+		return resp.data, resp.err
+	}
+}
+
+func (t *thumbnailHelper) pastelTopNodes(ctx context.Context, connections uint) (nodes node.List, err error) {
+	mns, err := t.pastelClient.MasterNodesTop(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for i, mn := range mns {
+		if uint(i) == connections {
+			break
+		}
+
+		nodes = append(nodes, node.NewNode(t.nodeClient, mn.ExtAddress, mn.ExtKey))
+	}
+
+	return nodes, nil
+}
+
+// Close closes request channel. Once Close is called, the helper can no longer fetch thumbnails
+func (t *thumbnailHelper) Close() {
+	if t.isClosed {
```

Comment:
why this check is outside of mutex locking section?

Reply:
fixed, good catch!